libass/ass.c

```diff
@@ -971,54 +965,122 @@ static int process_line(ASS_Track *track, char *str)
     return 0;
 }
 
-static int process_text(ASS_Track *track, char *str)
+static int process_text(ASS_Track *track, ASS_StringView *str)
 {
-    char *p = str;
     while (1) {
-        char *q;
         while (1) {
-            if ((*p == '\r') || (*p == '\n'))
-                ++p;
-            else if (p[0] == '\xef' && p[1] == '\xbb' && p[2] == '\xbf')
-                p += 3;         // U+FFFE (BOM)
+            if (str->len && (*str->str == '\r' || *str->str == '\n'))
+                ++str->str, --str->len;
+            else if (str->len >= 3 &&
+                     str->str[0] == '\xef' && str->str[1] == '\xbb' && str->str[2] == '\xbf')
+                str->str += 3, str->len -= 3; // U+FFFE (BOM)
             else
                 break;
         }
-        for (q = p; ((*q != '\0') && (*q != '\r') && (*q != '\n')); ++q) {
-        };
-        if (q == p)
+        ASS_StringView line = *str;
+        for (line.len = 0;
+             line.len < str->len && line.str[line.len] != '\r' && line.str[line.len] != '\n';
+             line.len++)
+             ;
+        if (!line.len || line.len == str->len)
             break;
-        if (*q != '\0')
-            *(q++) = '\0';
-        process_line(track, p);
-        if (*q == '\0')
-            break;
-        p = q;
+        process_line(track, line);
+        str->str += line.len, str->len -= line.len;
     }
     // there is no explicit end-of-font marker in ssa/ass
     if (track->parser_priv->fontname)
         decode_font(track);
     return 0;
 }
 
+static int process_text_full(ASS_Track *track, ASS_StringView str)
+{
+    int ret = process_text(track, &str);
+    if (ret < 0)
+        return ret;
+
+    if (!str.len)
+        return 0;
+
+    // Absurd? Yes.
+    if (str.len >= SIZE_MAX - 1)
+        return -ENOMEM;
+
+    char *buf = malloc(str.len + 1);
+    if (!buf)
+        return -ENOMEM;
+
+    str.str = buf;
+    str.len++;
+
+    ret = process_text(track, &str);
+
+    free(buf);
+    return ret;
+}
+
 /**
  * \brief Process a chunk of subtitle stream data.
  * \param track track
  * \param data string to parse
  * \param size length of data
 */
-void ass_process_data(ASS_Track *track, char *data, int size)
+void ass_process_data(ASS_Track *track, const char *data, int size)
 {
-    char *str = malloc(size + 1);
-    if (!str)
-        return;
+    ass_msg(track->library, MSGL_DBG2, "Event: %*s", size, data);
+
+    if (track->process_data_buf) {
+        int pos;
+        for (pos = 0; pos < size; pos++)
+            if (data[pos] == '\r' ||
+                data[pos] == '\n')
+                break;
+
+        if (pos == size) {
+            if (track->process_data_buf_size >= SIZE_MAX - size)
+                goto wipe;
 
-    memcpy(str, data, size);
-    str[size] = '\0';
+            char *newbuf = realloc(track->process_data_buf, track->process_data_buf_size + size);
+            if (!newbuf) {
+wipe:
```

Comment:
Better to move cleanup block at the end of the function.

Reply:
Sure.